lib/GADS.pm

```diff
@@ -3024,6 +3025,99 @@ prefix '/:layout_name' => sub {
+
+        if(param 'submit') {
+            my @columns = body_parameters->get_all('record');
+            $records->columns([@columns]);
+        } elsif(param 'purge') {
+            my @columns = split(/,/,body_parameters->get('records'));
+            $records->columns([@columns]);
+        }
+
+        my %ignores = (
+            id => 1,
+            createddate => 1,
+            createdby => 1,
+            serial => 1,
+        );
+
+        unless(param 'purge') {
+            while (my $record = $records->single) {
+                foreach my $column (@{$record->presentation->{columns}}) {
+                    next if $ignores{$column->{type}};
+                    my $result = +{};
+                    $result->{col_id} = $column->{id};
+                    $result->{name} = $column->{name};
+                    $result->{id} = $column->{data}->{record_id};
+                    $result->{value} = $column->{data}->{value};
+                    push @$record_data, $result;
+                }
+            }
+        }
+
+        $params->{page} = 'initial_purge';
+
+        if(param 'submit') {
+            $params->{page} = 'purge_stage_2';
+            $params->{records} = $record_data;
+            my @mapped_ids = map {$_->{col_id}} @$record_data;
+            my @mapped_records = uniq map {$_->{id}} @$record_data;
+            $params->{column_list} = join(',', @mapped_ids);
+            $params->{record_list} = join(',', @mapped_records);
+        } elsif(param 'purge') {
+            my $columns = body_parameters->get('columns') or error __"Unable to get columns";
+            my $rids = body_parameters->get('records') or error __"Unable to get records";
+            my @mapped_ids = split(/,/,$columns);
+            my @records = split(/,/,$rids);
+            my $total = 0;
+            foreach my $r (@records) {
+                my $record = schema->resultset('Record')->find($r);
+                my $current_id = $record->current_id;
+                my $current = schema->resultset('Current')->find($current_id);
+                $current->historic_purge(@mapped_ids);
+                $total++;
+            }
+            forwardHome({ success => "$total records have now been purged" }, $layout->identifier) if $total;
+            forwardHome({ danger => "No records were purged" }, $layout->identifier) unless $total;
+        } else {
+            my $result = [];
+            my $seen = {};
+            foreach my $val (@$record_data) {
+                if($seen->{$val->{name}}) {
+                    foreach my $d (@$result) {
+                        next unless $d->{name} eq $val->{name};
+                        $d->{count}++;
+                    }
+                }
+                else {
+                    $seen->{$val->{name}} = 1;
+                    push @$result, {id=>$val->{col_id}, name => $val->{name}, count => 1};
+                }
+            }
+            $params->{columns} = $result;
```

Comment:
I think there should be a way of removing most of the above and doing a simpler version in the template?